### PR TITLE
add torchmetrics wrapper for evaluate_boxes

### PR DESCRIPTION
**Pull request by jveitchmichaelis**
## Description

- Use `torchmetrics` to collect results and call `evaluate_boxes` during validation
- All metrics now respect `validate_on_epoch`, but should be fast enough even with large datasets to run with `n=1`. Metrics are always reset regardless.
- Simpler empty frame tracking during `validation_step`
- Much cleaner logging in `main`. Non-loggable metrics are dropped in the torchmetric class.
- Metric set up is delegated to when the trainer is created. See next point.
- If the user changes anything like validation data etc after initialization, we need to also re-init the metrics because precision/recall needs to know about label_dict and the annotation file when it calls `evaluate` under the hood. There are some guards in the code to check this (e.g. reload metrics in create_trainer and on model load).
- There is also a guard to not init metrics unless a validation set is defined, which is required for our internal metric (since it needs to know about the gt dataframe to run). Case when the user sets validation csv/root after init, instead of a config arg. I would suggest we think about decoupling this in evaluate for training (ie the core eval methods should only care about pairs of pred/target and not worry about image paths.
- The metric has a `task` arg which is set to `box` for now, as I anticipate that we will modify it to support different prediction types.
 
## Related Issue(s)

https://github.com/weecology/DeepForest/issues/901 (step towards this)
https://github.com/weecology/DeepForest/pull/1254 
https://github.com/weecology/DeepForest/issues/1245
Supports https://github.com/weecology/DeepForest/pull/1253

## AI-Assisted Development

- [X] I used AI tools (e.g., GitHub Copilot, ChatGPT, etc.) in developing this PR
- [X] I understand all the code I'm submitting
- [X] I have reviewed and validated all AI-generated code

**AI tools used (if applicable):**
Claude Code to do some bug hunting and experiments with gpu sync that turned out to be not required.

### CODE REVIEW

**Comment by bw4sz** [comment]
Is this ready for review? I'm seeing multi-gpu core dumps currently after 1 epoch, so i'm pretty sure its something about eval. Are you able to test this PR on 2 gpus? Not sure if its related, could be logging images, or logging predictions. Not sure which.

**Comment by bw4sz** [comment]
Let me see if a monkey merge it into train_birds just a moment, run that branch, which shouldn't have any changes to src and then see if it helps the core dump.

**Review by bw4sz — APPROVED** [review]
I am approving these changes with the belief that the multi_gpu errors i'm seeing in bird_training are unrelated. The fact that these tests pass is indication that we have improved the validation setup and it will help us with points and polygons.
